### ISSUE TITLE
A crash when working with a customized gate

### ISSUE THREAD

**Issue by yaelbh** [Qiskit/qiskit-terra#1635]
### Informations

- **Qiskit Terra version**:
- **Python version**:
- **Operating system**:

### What is the current behavior?

Trying to compile a circuit with a customized gate, I get:
```
Traceback (most recent call last):
  File "test_rb.py", line 50, in test_rb
    qobj = qiskit.compile(qc_test, backend=backend, basis_gates=basis_gates, shots=shots)
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/tools/compiler.py", line 61, in compile
    seed_mapper, pass_manager)
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/transpiler/_transpiler.py", line 67, in transpile
    'pass_manager': pass_manager})
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/tools/parallel.py", line 93, in parallel_map
    return [task(values[0], *task_args, **task_kwargs)]
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/transpiler/_transpiler.py", line 96, in _transpilation
    dag = circuit_to_dag(circuit)
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/converters/circuit_to_dag.py", line 85, in circuit_to_dag
    inst.cargs, control)
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/dagcircuit/_dagcircuit.py", line 364, in apply_operation_back
    self._check_basis_data(op, qargs, cargs)
  File "/opt/anaconda3/envs/YaelEnv/lib/python3.6/site-packages/qiskit/dagcircuit/_dagcircuit.py", line 260, in _check_basis_data
    raise DAGCircuitError("%s is not in the list of basis operations" % op.name)
qiskit.dagcircuit._dagcircuiterror.DAGCircuitError: 'r is not in the list of basis operations'
```

### Steps to reproduce the problem

Define the new gate:
```
class RGate(Gate):
    """R gate."""

    def __init__(self, qubit, circ=None):
        """Create new r gate."""
        super().__init__("r", [], [qubit], circ)

    def _define_decompositions(self):
        """
        gate r q { h q; s q; h q; s q; }
        """
        decomposition = DAGCircuit()
        q = QuantumRegister(1, "q")
        decomposition.add_qreg(q)
        decomposition.add_basis_element("h", 1, 0, 0)
        decomposition.add_basis_element("s", 1, 0, 0)
        rule = [
            HGate(q[0]),
            SGate(q[0]),
            HGate(q[0]),
            SGate(q[0])
        ]
        for inst in rule:
            decomposition.apply_operation_back(inst)
        self._decompositions = [decomposition]

    def inverse(self):
        """Invert this gate."""
        inv = RdgGate(self.qargs[0])
        self.circuit.data[-1] = inv  # replaces the gate with the inverse
        return inv

    def reapply(self, circ):
        """Reapply this gate to corresponding qubits in circ."""
        self._modifiers(circ.r(self.qargs[0]))

def r(self, q):
    """Apply R to q."""

    if isinstance(q, QuantumRegister):
        instructions = InstructionSet()
        for j in range(q.size):
            instructions.add(self.r((q, j)))
        return instructions

    self._check_qubit(q)
    return self._attach(RGate(q, self))

QuantumCircuit.r = r
```

Then write some circuit with the new gate, and compile:
```
qobj = qiskit.compile(qc_test, backend=backend, basis_gates=basis_gates, shots=shots)
```

### What is the expected behavior?

Do not crash. Perhaps I'm simply doing something wrong?

**Comment by Exferro** [comment]
@yaelbh Hi!

Briefly: you'd have to use CompositeGate class.

In more details: I am not aware of current situation (Qiskit 0.7), but it was a known problem for Qiskit 0.6. Here is the answer I obtained in the Slack channel:

>ad_su [2 months ago]
I don't really have time to explain everything right now but you have several examples of user defined quantum gates here: https://github.com/nelimee/quantum-tools/tree/master/gates .
Be warned that the CompositeGate API is known to be unintuitive and that it will eventually change (but we don't know when, it's been 4 months the API is in this state).

You can find more information in  #1230 and others similar issues.

Hope, it helped you!

**Comment by yaelbh** [comment]
Thank you!

**Comment by 1ucian0** [comment]
I think this can be closed via #1816. Feel free to reopen if that's not the case.

> CompositeGate is deprecated and will be removed in v0.9.
> Any Instruction can now be composed of other sub-instructions.
> To build them, you construct a circuit then use
> circuit.to_instruction().
